src/cscapi/client.py

```diff
@@ -0,0 +1,276 @@
+                except httpx.HTTPStatusError as exc:
+                    logging.error(
+                        f"Error while sending metrics: {exc} for machine {machine_to_process.machine_id}"
+                    )
+
+            attempt_count += 1
+            machines_to_process_attempts = next_machines_to_process_attempts
+
+    def _send_signals(self, token: str, signals: SignalModel):
+        for signal_batch in batched(signals, 250):
+            body = [asdict(signal) for signal in signal_batch]
+            resp = self.http_client.post(
+                CAPI_SIGNALS_URL, json=body, headers={"Authorization": token}
+            )
+            resp.raise_for_status()
+
+    def _send_metrics_for_machine(self, machine: MachineModel):
+        resp = self.http_client.post(
+            CAPI_METRICS_URL,
+            json={
+                "bouncers": [],
+                "machines": [
+                    {
+                        "last_update": datetime.datetime.now().isoformat(),
+                        "last_push": datetime.datetime.now().isoformat(),
+                        "version": __version__,
+                        "name": machine.machine_id,
+                    }
+                ],
+            },
+        )
+        resp.raise_for_status()
+
+    def _prune_signals(self):
+        signals = filter(lambda signal: signal.sent, self.storage.get_all_signals())
+        self.storage.delete_signals(signals)
+
+    def _refresh_machine_token(self, machine: MachineModel) -> MachineModel:
+        resp = self.http_client.post(
+            CAPI_WATCHER_LOGIN_URL,
+            json={
+                "machine_id": machine.machine_id,
+                "password": machine.password,
+                "scenarios": machine.scenarios.split(","),
+            },
+        )
+        try:
+            resp.raise_for_status()
+        except httpx.HTTPStatusError as exc:
+            logging.error(
+                f"Error while refreshing token: machine_id might be already registered or password is wrong"
+            )
+            raise exc
+
+        new_machine = asdict(machine)
+        new_machine["token"] = resp.json()["token"]
+        new_machine = MachineModel(**new_machine)
+        self.storage.update_or_create_machine(new_machine)
+        return new_machine
+
+    def _register_machine(self, machine: MachineModel) -> MachineModel:
+        logging.info(f"registering machine {machine.machine_id}")
+        machine.password = (
+            machine.password if machine.password else secrets.token_urlsafe(32)
+        )
+        resp = self.http_client.post(
+            CAPI_WATCHER_REGISTER_URL,
+            json={
+                "machine_id": machine.machine_id,
+                "password": machine.password,
+            },
+        )
+        self.storage.update_or_create_machine(machine)
+        return machine
+
+    def _make_machine(self, machine: MachineModel):
```

Comment:
[warn] we should call this in the loop of while machines_to_process_attempts
We can assume that it should not fail so this would run most of the time once per machine anyways.
And you ensure the machine is freshly created hence token still valid as you handle the machine.
Also, if we sleep between attempt (we should) the refresh is really fresh

Reply:
Sorry I don't get this. 

This method is for https://crowdsecurity.github.io/api_doc/index.html?urls.primaryName=CAPI#/watchers/post_watchers

The token is created in another method. 

All this dumb method does is create random password for machine and registers it. Not the token.

Also didn't get which attempt you're referring to.

Reply:
referring to the _refresh_machine_token that is call in this function and it might be too early. it's better to do it as close as possible to the code about to do the curl call (begining of the send signal for a machine)